mtp_noms_ops/apps/security/views/object_base.py

```diff
@@ -112,35 +139,66 @@ def form_valid(self, form):
+        ):
+            search_results_url = f'{reverse(self.search_results_view)}?{form.query_string}'
+            return redirect(search_results_url)
+
         context = self.get_context_data(form=form)
         object_list = form.get_object_list()
         form.check_and_update_saved_searches(self.title)
         if self.redirect_on_single and len(object_list) == 1 and hasattr(self, 'url_for_single_result'):
             return redirect(self.url_for_single_result(object_list[0]))
         context[self.object_list_context_key] = object_list
-        return render(self.request, self.template_name, context)
+        return render(self.request, self.get_template_names(), context)
 
     def form_invalid(self, form):
-        if self.export_view:
-            return self.get_export_redirect(form)
+        """
+        If the form is invalid and the view type is export or search results,
+        it redirects back to the referral view so that the user can see and correct the errors.
+        """
+        if self.is_export_view_type() or self.view_type == ViewType.search_results:
+            return self.redirect_to_referral_view(form)
         return super().form_invalid(form)
 
     def get_context_data(self, **kwargs):
         context_data = super().get_context_data(**kwargs)
-        context_data['breadcrumbs'] = [
-            {'name': _('Home'), 'url': reverse('security:dashboard')},
-            {'name': self.title}
-        ]
-        context_data['google_analytics_pageview'] = genericised_pageview(self.request, self.get_generic_title())
-        return context_data
 
-    def get_export_redirect(self, form):
-        return redirect('%s?%s' % (reverse(self.export_redirect_view, kwargs=self.kwargs), form.query_string))
+        if self.view_type == ViewType.search_results:
+            breadcrumbs = [
+                {'name': _('Home'), 'url': reverse('security:dashboard')},
+                {'name': self.title, 'url': f'{reverse("security:sender_list")}?{kwargs["form"].query_string}'},
+                {'name': _('Search results')}
+            ]
+        else:
+            breadcrumbs = [
+                {'name': _('Home'), 'url': reverse('security:dashboard')},
+                {'name': self.title}
+            ]
+
+        return {
+            **context_data,
+
+            'breadcrumbs': breadcrumbs,
+            'google_analytics_pageview': genericised_pageview(self.request, self.get_generic_title()),
+            'simple_search_form_submitted_input_name': SIMPLE_SEARCH_FORM_SUBMITTED_INPUT_NAME,
+            'is_search_results': self.view_type == ViewType.search_results,
+        }
+
+    def redirect_to_referral_view(self, form):
+        """
+        Returns an HttpResponseRedirect to the referral_view preserving the same kwargs and query string.
+        """
+        view = reverse(self.referral_view, kwargs=self.kwargs)
+        return redirect(f'{view}?{self.request.GET.urlencode()}')
```

Comment:
quick question, why us this using GET parameters rather than `form.query_string` like before? just simplicity or is there something in GET that's not in the form?

Reply:
that's because if the form is invalid query_string won't have the correct params, it's better to preserve the invalid params and use request.GET instead I think.